### PR TITLE
Save the workspace list across restarts

### PR DESCRIPTION
**Pull request by billti**

### CODE REVIEW

File: vscode/src/azure/commands.ts [hunk 1]
```diff
@@ -21,13 +22,27 @@ import { QuantumUris, compileToBitcode } from "./networkRequests";
 import { getQirForActiveWindow } from "../qirGeneration";
 
 const corsDocsUri = "https://github.com/microsoft/qsharp/wiki/Enabling-CORS";
+const workspacesSecret = "qsharp-vscode.workspaces";
 
-export function initAzureWorkspaces() {
+export async function initAzureWorkspaces(context: vscode.ExtensionContext) {
```
Comment (idavis) [inline vscode/src/azure/commands.ts:27]: Should we add the return type now that this is async?

File: vscode/src/azure/commands.ts [hunk 2]
```diff
@@ -21,13 +22,27 @@ import { QuantumUris, compileToBitcode } from "./networkRequests";
...
   vscode.window.createTreeView("quantum-workspaces", {
     treeDataProvider: workspaceTreeProvider,
   });
 
+  // Add any previously saved workspaces
+  const savedWorkspaces = await context.secrets.get(workspacesSecret);
+  if (savedWorkspaces) {
+    log.debug("Loading workspaces: ", savedWorkspaces);
+    const workspaces: WorkspaceConnection[] = JSON.parse(savedWorkspaces);
+    for (const workspace of workspaces) {
```
Comment (idavis) [inline vscode/src/azure/commands.ts:38]: Should we use `Promise.all()`?

File: vscode/src/azure/commands.ts [hunk 3]
```diff
@@ -75,25 +90,57 @@ export function initAzureWorkspaces() {
         return;
       }
 
-      setTimeout(async () => {
-        await queryWorkspace(arg.workspace);
-        workspaceTreeProvider.updateWorkspace(arg.workspace);
-      }, 1000);
+      setTimeout(() => {
+        workspaceTreeProvider.refresh();
+      }, 2000);
```
Comment (idavis) [inline vscode/src/azure/commands.ts:95]: nit: magic numbers should be const vars.

**Review by minestarks — APPROVED** [review]

**Review by idavis — APPROVED** [review]

File: vscode/src/azure/commands.ts (hunk 1 above)
Comment (billti) [reply vscode/src/azure/commands.ts:27]: It should still be inferred correctly, as it was before.

File: vscode/src/azure/commands.ts (hunk 2 above)
Comment (billti) [reply vscode/src/azure/commands.ts:38]: I actually re-wrote this a couple times after hitting some async hazards, and this was the safest and simplest code. There are some subtle bugs and gotchas when you iterate through a list and kick of async work for each item. (I expect the vast majority of users will only have the one workspace in the tree anyway).

If you have a strong opinion though, I could be convinced.

File: vscode/src/azure/commands.ts [hunk 4]
```diff
@@ -75,25 +90,57 @@
         return;
       }
 
-      setTimeout(async () => {
-        await queryWorkspace(arg.workspace);
-        workspaceTreeProvider.updateWorkspace(arg.workspace);
-      }, 1000);
+      setTimeout(() => {
+        workspaceTreeProvider.refresh();
```
[the remark this answers is not in the document]
Comment (billti) [reply vscode/src/azure/commands.ts:94]: It doesn't handle untrusted code.

**Pull request merged**